### PR TITLE
Python dependency refactor

### PR DESCRIPTION
**Pull request by eli-schwartz**
This moves the python dependency factory out of modules/python.py and into dependencies/python.py. As a result, we get some consistency -- particularly, bugfixes! for the sysconfig resolver for `dependency('python3')`. It's also a lot cleaner, and easier to read modules/.

Finally, fix bad overlinking in the sysconfig dependency.

### CODE REVIEW

File: mesonbuild/modules/python.py [hunk 1]
```diff
@@ -428,6 +431,7 @@ def __init__(self, name: str, command: T.Optional[T.List[str]] = None,
             'sysconfig_paths': {},
             'paths': {},
             'platform': 'sentinal',
```
Comment (dnicolodi) [inline mesonbuild/modules/python.py]: Not part of the patch, but this looks like a typo.

File: mesonbuild/dependencies/python.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+from .base import DependencyMethods, SystemDependency
+from .factory import DependencyFactory
+from ..environment import detect_cpu_family
+
+if T.TYPE_CHECKING:
+    from ..environment import Environment
+
+
+class Python3DependencySystem(SystemDependency):
+    def __init__(self, name: str, environment: 'Environment', kwargs: T.Dict[str, T.Any]) -> None:
```
Comment (dnicolodi) [inline mesonbuild/dependencies/python.py]: I see that you are just moving code around and I don't know what the the Meson coding style says about this, but if you have `from __futute__ import annotations` using string for the annotations is redundant as they are not resolved when not type checking anyway. Namely, this could be `environment: Environment`.

File: mesonbuild/dependencies/python.py [hunk 3]
```diff
@@ -170,11 +170,24 @@ def __init__(self, name: str, environment: 'Environment',
...
+            self.variables.get('INCLUDEPY'),
+            self.paths.get('include'),
+            self.paths.get('platinclude')])
+
+        self.compile_args += ['-I' + path for path in inc_paths if path]
+
+        # https://sourceforge.net/p/mingw-w64/mailman/message/30504611/
+        if mesonlib.is_windows() and self.get_windows_python_arch() == '64' and self.major_version == 2:
+            self.compile_args += ['-DMS_WIN64']
+
```
Comment (dnicolodi) [inline mesonbuild/dependencies/python.py:193]: Again, not related to the refactoring, but compiling Cython extension modules with GCC on Windows requires to add this to `meson.build` also for Python 3:
```meson
if host_machine.system() == 'windows' and meson.get_compiler('c').get_id() == 'gcc'
  add_project_arguments('-DMS_WIN64', language: ['c', 'cpp'])
endif
```
therefore I think the `and self.major_version == 2` guard should be dropped.

**Comment by dnicolodi** [comment]
By the way, how does this interact with #11133? I see that both PRs contain a commit to move the introspection script to a separate file. It would be great to have #11133 merged soon too.

File: mesonbuild/dependencies/python.py (hunk 3 above)
Comment (eli-schwartz) [reply mesonbuild/dependencies/python.py:193]: I really have no clue how that works, but I guess that can be done... in a separate commit. ;)

@MathieuDuponchelle, do you happen to remember why you originally guarded it by version 2.x, way back in commit e1b138a21b45fe9b55f3bb7d9b1e0bbbd45a7609?

**Comment by eli-schwartz** [comment]
A couple of base commits are shared in common with both PRs, yes. As you can see, there's a bunch of cleanups here I'm trying to do. :D

As far as merging it goes, I still have both my PRs to CPython open, and I was sort of hoping to get confirmation that that's what we're supposed to do before merging that same logic but ported to our introspection script. I haven't heard back yet, though.

File: mesonbuild/dependencies/python.py (hunk 3 above)
Comment (MathieuDuponchelle) [reply mesonbuild/dependencies/python.py:193]: Hey, I should have documented that part specifically but only added the thread that justified adding `-DMS_WIN64`, I have absolutely no clue at this point why I added the python 2 restriction, sorry about that.

Comment (eli-schwartz) [reply mesonbuild/dependencies/python.py:193]: No big deal, looks like it may have just been due to unclear wording in the linked message...

I bet @rgommers knows a lot about the topic anyway and can confirm the correctness of such a change.

Comment (rgommers) [reply mesonbuild/dependencies/python.py:193]: The SciPy code comment points at https://github.com/python/cpython/issues/72454. There's a bunch of MinGW patches that are needed but were never accepted by CPython (they basically never cared much about anything but MSVC). It looks like it was very recently resubmitted and merged (https://github.com/python/cpython/pull/100137), so it should be fixed for Python `3.12.0`. So the define is needed for Python 3 up to today, and hence dropping the `== 2` sounds good to me.

And, just to be sure: if `-DMS_WIN64` is added twice (once by Meson, once by e.g. SciPy), that won't hurt will it?

Comment (eli-schwartz) [reply mesonbuild/dependencies/python.py:193]: Thanks for the context -- that's wonderfully actionable and gives me a lot of confidence about what to do here.

And yes, double-defining it should be fine, the definitions are compatible so it's a no-op.

File: mesonbuild/dependencies/python.py [hunk 4]
```diff
@@ -13,18 +13,95 @@
 # limitations under the License.
 from __future__ import annotations
 
+import json, sysconfig
```
Comment (dcbaker) [inline mesonbuild/dependencies/python.py]: Can we please keep these split into:
```python
import json
import sysconfig
```

File: mesonbuild/dependencies/python.py [hunk 5]
```diff
@@ -13,18 +13,95 @@
...
 from pathlib import Path
-import sysconfig
 import typing as T
 
-from .. import mlog
+from .. import mesonlib, mlog
```
Comment (dcbaker) [inline mesonbuild/dependencies/python.py:20]: and the same here.

**Review by dcbaker — APPROVED** [review]
I've got a few style nits, but otherwise this looks good to me.

File: mesonbuild/dependencies/python.py (hunk 5 above)
Comment (eli-schwartz) [reply mesonbuild/dependencies/python.py:20]: The thing is... immediately below, we do exactly the same thing with e.g.

```diff
+from .base import process_method_kw, DependencyMethods, DependencyTypeName, ExternalDependency, SystemDependency
```

I really find it more natural and idiomatic to do this, and I don't see the point in spreading imports across one bound name per line. That just becomes difficult to read and difficult to scroll past.

Comment (dcbaker) [reply mesonbuild/dependencies/python.py:20]: Idiomatically modules are imported one per line, objects from modules are imported on a single line:
```python
import os
import sys

from os.path import join, dirname, basename
```
https://peps.python.org/pep-0008/#imports

I get it, style guides always have something in them that annoys people. But having a single styl, makes the whole codebase much more readable.

Comment (eli-schwartz) [reply mesonbuild/dependencies/python.py:20]: And this is `from foo import bar, baz`...

I'm really unhappy that people keep campaigning to make ugly code, the same people who defend "modules are imported one per line" are defending "objects from modules are imported one per line" and for the same reason.

This is nothing to do with style. I've already stated my ***strong*** objections to "add a lint rule to force objects from modules to be imported one per line because I think it's better for rebasing". What I actually want is a lint rule that forbids that.
